storage: Improve validation of split key

bdarnell opened the pull request:
Use the descriptor version that was passed in instead of r.Desc(), to
ensure that the decision is consistent with the descriptor as used
elsewhere in the method.

Validate that the key is in-bounds after all
transformations (including EnsureSafeSplitKey) have been completed.

May detect issues like those hypothesized in #16008

bdarnell committed:
storage: Improve validation of split key

Use the descriptor version that was passed in instead of r.Desc(), to
ensure that the decision is consistent with the descriptor as used
elsewhere in the method.

Validate that the key is in-bounds after all
transformations (including EnsureSafeSplitKey) have been completed.

May detect issues like those hypothesized in #16008

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/16291)
<!-- Reviewable:end -->

tschottdorf commented:
LGTM. A test would be nice, but I think this should merge ASAP.

---

Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16291#-:-Kldf5_r5ixAj2xxyFYv:bc0e8mv)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Added a test, which showed that the original change didn't work as expected. Returning RangeKeyMismatchError here results in an infinite loop. I put the RangeKeyMismatchError check back where it was, and added a *second* validation step that returns a non-retryable error if the key was originally valid but changed by EnsureSafeSplitKey.

tschottdorf commented:
Yikes, good catch. Thanks for adding the test. <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16291#-:-Kle3ZPUgQHpgY8B7Adp:b-du7m1e)*
<!-- Sent from Reviewable.io -->
